crates/net/eth-wire/src/multiplex.rs

```diff
@@ -0,0 +1,459 @@
+        self.from_wire.poll_next_unpin(cx)
+    }
+}
+
+/// A Stream and Sink type that acts as a wrapper around a primary RLPx subprotocol (e.g. "eth")
+/// [EthStream](crate::EthStream) and can also handle additional subprotocols.
+#[derive(Debug)]
+pub struct RlpxSatelliteStream<St, Primary> {
+    /// The raw p2p stream
+    conn: P2PStream<St>,
+    to_primary: UnboundedSender<BytesMut>,
+    from_primary: UnboundedReceiverStream<Bytes>,
+    primary: Primary,
+    primary_capability: SharedCapability,
+    satellites: Vec<ProtocolStream>,
+    out_buffer: VecDeque<Bytes>,
+}
+
+impl<St, Primary> RlpxSatelliteStream<St, Primary> {}
+
+impl<St, Primary, PrimaryErr> Stream for RlpxSatelliteStream<St, Primary>
+where
+    St: Stream<Item = io::Result<BytesMut>> + Sink<Bytes, Error = io::Error> + Unpin,
+    Primary: TryStream<Error = PrimaryErr> + Unpin,
+    P2PStreamError: Into<PrimaryErr>,
+{
+    type Item = Result<Primary::Ok, Primary::Error>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+
+        loop {
+            // first drain the primary stream
+            if let Poll::Ready(Some(msg)) = this.primary.try_poll_next_unpin(cx) {
+                return Poll::Ready(Some(msg))
+            }
+
+            let mut out_ready = true;
+            loop {
+                match this.conn.poll_ready_unpin(cx) {
+                    Poll::Ready(_) => {
+                        if let Some(msg) = this.out_buffer.pop_front() {
+                            if let Err(err) = this.conn.start_send_unpin(msg) {
+                                return Poll::Ready(Some(Err(err.into())))
+                            }
+                        } else {
+                            break;
+                        }
+                    }
+                    Poll::Pending => {
+                        out_ready = false;
+                        break
+                    }
+                }
+            }
+
+            // advance primary out
+            loop {
+                match this.from_primary.poll_next_unpin(cx) {
+                    Poll::Ready(Some(msg)) => {
+                        this.out_buffer.push_back(msg);
+                    }
+                    Poll::Ready(None) => {
+                        // primary closed
+                        return Poll::Ready(None)
+                    }
+                    Poll::Pending => break,
+                }
+            }
+
+            // advance all satellites
+            for idx in (0..this.satellites.len()).rev() {
+                let mut proto = this.satellites.swap_remove(idx);
+                loop {
+                    match proto.poll_next_unpin(cx) {
+                        Poll::Ready(Some(msg)) => {
+                            this.out_buffer.push_back(msg);
+                        }
+                        Poll::Ready(None) => return Poll::Ready(None),
```

Comment:
This was confusing to me at first, but if _any_ of the streams are closed, we
need to return `Poll::Ready(None)`